community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/ImportLogic.java

```diff
@@ -0,0 +1,528 @@
+                nodeRelationshipCache, readFilter, denseChangeFilter, nodeTypes ) );
+
+        updatePeakMemoryUsage();
+
+        if ( upToType == relationshipTypeDistribution.getNumberOfRelationshipTypes() )
+        {
+            // This means that we've linked all the types
+            nodeRelationshipCache.close();
+            nodeRelationshipCache = null;
+            return -1;
+        }
+
+        return upToType;
+    }
+
+    /**
+     * Convenience method (for code reading) to have a zero-based value become one based (for printing/logging).
+     */
+    private static int oneBased( int value )
+    {
+        return value + 1;
+    }
+
+    /**
+     * @return index (into {@link RelationshipTypeDistribution}) of last relationship type that fit in memory this round.
+     */
+    static int nextSetOfTypesThatFitInMemory( RelationshipTypeDistribution typeDistribution, int startingFromType,
+            long freeMemoryForDenseNodeCache, long numberOfDenseNodes )
+    {
+        assert startingFromType >= 0 : startingFromType;
+
+        long currentSetOfRelationshipsMemoryUsage = 0;
+        int numberOfTypes = typeDistribution.getNumberOfRelationshipTypes();
+        int toType = startingFromType;
+        for ( ; toType < numberOfTypes; toType++ )
+        {
+            // Calculate worst-case scenario
+            Pair<Object,Long> type = typeDistribution.get( toType );
+            long relationshipCountForThisType = type.other();
+            long memoryUsageForThisType = calculateMaxMemoryUsage( numberOfDenseNodes, relationshipCountForThisType );
+            long memoryUsageUpToAndIncludingThisType =
+                    currentSetOfRelationshipsMemoryUsage + memoryUsageForThisType;
+            if ( memoryUsageUpToAndIncludingThisType > freeMemoryForDenseNodeCache &&
+                    currentSetOfRelationshipsMemoryUsage > 0 )
+            {
+                // OK the current set of types is enough to fill the cache
+                break;
+            }
+
+            currentSetOfRelationshipsMemoryUsage += memoryUsageForThisType;
+        }
+        return toType;
+    }
+
+    /**
+     * Optimizes the relationship groups store by physically locating groups for each node together.
+     */
+    public void defragmentRelationshipGroups()
+    {
+        // Defragment relationships groups for better performance
+        new RelationshipGroupDefragmenter( config, executionMonitor, numberArrayFactory )
+                .run( max( maxMemory, peakMemoryUsage ), neoStore, neoStore.getNodeStore().getHighId() );
+    }
+
+    /**
+     * Builds the counts store. Requires that {@link #importNodes()} and {@link #importRelationships()} has run.
+     */
+    public void buildCountsStore()
+    {
+        // Count nodes per label and labels per node
+        try ( CountsAccessor.Updater countsUpdater = neoStore.getCountsStore().reset(
```

Comment:
we have a CountsComputer thingy that does kind of same thing already

Reply:
Yes, kind of... it has other means of progress and configuration and what not. Do you feel it's absolutely necessary to do this change? It's also a bit unrelated to this PR

Reply:
it fine if we will fix it in future PR, we will need to do that anyway when we will extract importer out of kernel